File: src/cc/redpen/intellij/SettingsPane.java
package cc.redpen.intellij;

import cc.redpen.RedPenException;
import cc.redpen.config.*;
import com.intellij.openapi.ui.Messages;
import com.intellij.ui.PopupMenuListenerAdapter;

import javax.swing.*;
import javax.swing.event.PopupMenuEvent;
import javax.swing.filechooser.FileNameExtensionFilter;
import javax.swing.table.DefaultTableModel;
import javax.swing.table.TableModel;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.*;
import java.util.stream.Stream;

import static java.util.stream.Collectors.toList;
import static java.util.stream.IntStream.range;
import static javax.swing.JFileChooser.APPROVE_OPTION;

public class SettingsPane {
  RedPenProvider provider;
  private Map<String, Configuration> configs = new LinkedHashMap<>();
  JPanel root;
  private JTabbedPane tabbedPane;
  JTable validators;
  JTable symbols;
  JComboBox<String> language;
  JCheckBox autodetectLanguage;
  JButton exportButton;
  JButton importButton;
  JButton resetButton;
  JFileChooser fileChooser = new JFileChooser();
  ConfigurationExporter configurationExporter = new ConfigurationExporter();
  ConfigurationLoader configurationLoader = new ConfigurationLoader();

  public SettingsPane(RedPenProvider provider) {
    this.provider = provider;
    cloneConfigs();
    fileChooser.setFileFilter(new FileNameExtensionFilter("RedPen Configuration", "xml"));
  }

  void cloneConfigs() {
    provider.getConfigs().forEach((k, v) -> configs.put(k, v.clone()));
  }

  public JPanel getPane() {
    initLanguages();
    initTabs();
    initButtons();
    return root;
  }

  void initButtons() {
    exportButton.addActionListener(a -> exportConfig());
    importButton.addActionListener(a -> importConfig());
    resetButton.addActionListener(a -> resetToDefaults());
  }

  void importConfig() {
    try {
      if (fileChooser.showOpenDialog(root) != APPROVE_OPTION) return;
      setConfig(configurationLoader.load(fileChooser.getSelectedFile()));
      initTabs();
    }
    catch (RedPenException e) {
      Messages.showMessageDialog("Cannot load: " + e.getMessage(), "RedPen", Messages.getErrorIcon());
    }
  }

  void exportConfig() {
    try {
      if (fileChooser.showSaveDialog(root) != APPROVE_OPTION) return;
      save();
      configurationExporter.export(getConfig(), new FileOutputStream(fileChooser.getSelectedFile()));
    }
    catch (IOException e) {
      Messages.showMessageDialog("Cannot write to file: " + e.getMessage(), "RedPen", Messages.getErrorIcon());
    }
  }

  void initLanguages() {
    provider.getConfigs().keySet().forEach(k -> language.addItem(k));
    autodetectLanguage.setSelected(provider.getAutodetect());
    language.setSelectedItem(provider.getActiveConfig().getKey());
    language.addPopupMenuListener(new PopupMenuListenerAdapter() {
      @Override public void popupMenuWillBecomeVisible(PopupMenuEvent e) {
        applyChanges();
      }
    });
    language.addActionListener(a -> initTabs());
  }

  void initTabs() {
    initValidators();
    initSymbols();
  }

  void initSymbols() {
    symbols.removeAll();
    DefaultTableModel model = createSymbolsModel();
    symbols.setModel(model);
    symbols.setRowHeight((int)(validators.getFont().getSize() * 1.5));

    model.addColumn("Symbols");
    model.addColumn("Value");
    model.addColumn("Invalid chars");
    model.addColumn("Space before");
    model.addColumn("Space after");

    symbols.getColumnModel().getColumn(0).setMinWidth(250);
    symbols.setDefaultEditor(Character.class, new SingleCharEditor());

    SymbolTable symbolTable = getConfig().getSymbolTable();
    for (SymbolType key : symbolTable.getNames()) {
      Symbol symbol = symbolTable.getSymbol(key);
      model.addRow(new Object[] {symbol.getType().toString(), symbol.getValue(),
        new String(symbol.getInvalidChars()), symbol.isNeedBeforeSpace(), symbol.isNeedAfterSpace()});
    }

    symbols.doLayout();
  }

  void initValidators() {
    validators.removeAll();
    DefaultTableModel model = createValidatorsModel();
    validators.setModel(model);
    validators.setRowHeight((int)(validators.getFont().getSize() * 1.5));

    model.addColumn("");
    model.addColumn("Name");
    model.addColumn("Properties (comma-separated)");

    validators.getColumnModel().getColumn(0).setMaxWidth(20);

    for (ValidatorConfiguration initialValidator : provider.getInitialConfig(getConfig().getKey()).getValidatorConfigs()) {
      Optional<ValidatorConfiguration> validator = getConfig().getValidatorConfigs().stream().filter(v -> v.equals(initialValidator)).findFirst();
      model.addRow(new Object[] { validator.isPresent(), initialValidator.getConfigurationName(), attributes(validator.orElse(initialValidator))});
    }

    validators.doLayout();
  }

  public List<ValidatorConfiguration> getActiveValidators() {
    List<ValidatorConfiguration> result = new ArrayList<>();
    if (validators.isEditing()) validators.getCellEditor().stopCellEditing();
    TableModel model = validators.getModel();
    for (int i = 0; i < model.getRowCount(); i++) {
       if ((boolean)model.getValueAt(i, 0)) {
         ValidatorConfiguration validator = provider.getInitialConfig(getConfig().getKey()).getValidatorConfigs().get(i).clone();
         validator.getAttributes().clear();
         String attributes = (String)model.getValueAt(i, 2);
         Stream.of(attributes.trim().split("\\s*,\\s*")).filter(s -> !s.isEmpty()).forEach(s -> {
           String[] attr = s.split("=", 2);
           if (attr.length < 2 || attr[0].isEmpty())
             showPropertyError(validator, s);
           else
             validator.addAttribute(attr[0], attr[1]);
         });
         result.add(validator);
       }
    }
    return result;
  }

  void showPropertyError(ValidatorConfiguration validator, String s) {
    Messages.showMessageDialog("Validator property must be in key=value format: " + s, validator.getConfigurationName(), Messages.getErrorIcon());
  }

  public List<Symbol> getSymbols() {
    if (symbols.isEditing()) symbols.getCellEditor().stopCellEditing();

    TableModel model = symbols.getModel();
    return range(0, model.getRowCount()).mapToObj(i -> new Symbol(
      SymbolType.valueOf((String)model.getValueAt(i, 0)), String.valueOf(model.getValueAt(i, 1)).charAt(0), (String)model.getValueAt(i, 2),
      (boolean)model.getValueAt(i, 3), (boolean)model.getValueAt(i, 4)
    )).collect(toList());
  }

  private String attributes(ValidatorConfiguration validatorConfig) {
    String result = validatorConfig.getAttributes().toString();
    return result.substring(1, result.length() - 1);
  }

  void applyValidatorsChanges() {
    List<ValidatorConfiguration> validators = getConfig().getValidatorConfigs();
    List<ValidatorConfiguration> remainingValidators = getActiveValidators();
    validators.clear();
    validators.addAll(remainingValidators);
  }

  void applySymbolsChanges() {
    SymbolTable symbolTable = getConfig().getSymbolTable();
    getSymbols().stream().forEach(symbolTable::overrideSymbol);
  }

  void applyChanges() {
    applyValidatorsChanges();
    applySymbolsChanges();
  }

  public void save() {
    applyChanges();
    provider.getConfigs().putAll(configs);
    cloneConfigs();
  }

  public void resetChanges() {
    cloneConfigs();
    initTabs();
  }

  public void resetToDefaults() {
    provider.getInitialConfigs().forEach((k, v) -> configs.put(k, v.clone()));
    initTabs();
  }

  DefaultTableModel createValidatorsModel() {
    return new DefaultTableModel() {
      @Override public Class<?> getColumnClass(int i) {
        return i == 0 ? Boolean.class : String.class;
      }

      @Override public boolean isCellEditable(int row, int column) {
        return column != 1;
      }
    };
  }

  DefaultTableModel createSymbolsModel() {
    return new DefaultTableModel() {
      @Override public Class<?> getColumnClass(int i) {
        return i == 1 ? Character.class : i == 3 || i == 4 ? Boolean.class : String.class;
      }

      @Override public boolean isCellEditable(int row, int column) {
        return column != 0;
      }
    };
  }

  public Configuration getConfig() {
    return getConfig((String)language.getSelectedItem());
  }

  public Configuration getConfig(String key) {
    return configs.get(key);
  }

  public void setConfig(Configuration config) {
    configs.put(config.getKey(), config);
    language.setSelectedItem(config.getKey());
    if (!config.getKey().equals(language.getSelectedItem())) {
      provider.addConfig(config);
      language.addItem(config.getKey());
      language.setSelectedItem(config.getKey());
    }
  }
}

File: src/cc/redpen/intellij/SettingsPane.kt
package cc.redpen.intellij

import cc.redpen.RedPenException
import cc.redpen.config.*
import com.intellij.configurationStore.save
import com.intellij.openapi.ui.Messages
import com.intellij.ui.PopupMenuListenerAdapter
import com.intellij.uiDesigner.core.GridConstraints
import com.intellij.uiDesigner.core.GridConstraints.*
import com.intellij.uiDesigner.core.GridLayoutManager
import com.intellij.uiDesigner.core.Spacer
import java.io.FileOutputStream
import java.io.IOException
import java.util.*
import javax.swing.*
import javax.swing.JFileChooser.APPROVE_OPTION
import javax.swing.event.CellEditorListener
import javax.swing.event.ChangeEvent
import javax.swing.event.PopupMenuEvent
import javax.swing.filechooser.FileNameExtensionFilter
import javax.swing.table.DefaultTableModel

open class SettingsPane(internal var provider: RedPenProvider) {
    open val configs: MutableMap<String, Configuration> = LinkedHashMap()
    internal var root = JPanel()
    internal var validators = JTable(createValidatorsModel())
    internal var symbols = JTable(createSymbolsModel())
    internal var language = JComboBox<String>()
    internal var exportButton = JButton("Export...")
    internal var importButton = JButton("Import...")
    internal var resetButton = JButton("Reset to defaults")
    internal var fileChooser = JFileChooser()
    internal var configurationExporter = ConfigurationExporter()
    internal var configurationLoader = ConfigurationLoader()

    init {
        cloneConfigs()
        fileChooser.fileFilter = FileNameExtensionFilter("RedPen Configuration", "xml")

        root.layout = GridLayoutManager(2, 7)
        root.add(JLabel("Language"), GridConstraints(0, 0, 1, 1, ANCHOR_WEST, FILL_NONE, SIZEPOLICY_FIXED, SIZEPOLICY_FIXED, null, null, null))
        root.add(language, GridConstraints(0, 1, 1, 1, ANCHOR_WEST, FILL_HORIZONTAL, SIZEPOLICY_FIXED, SIZEPOLICY_FIXED, null, null, null))
        root.add(Spacer(), GridConstraints(0, 3, 1, 1, ANCHOR_CENTER, FILL_HORIZONTAL, SIZEPOLICY_WANT_GROW, SIZEPOLICY_CAN_SHRINK, null, null, null))
        root.add(exportButton, GridConstraints(0, 5, 1, 1, ANCHOR_CENTER, FILL_HORIZONTAL, SIZEPOLICY_CAN_SHRINK or SIZEPOLICY_CAN_GROW, SIZEPOLICY_FIXED, null, null, null))
        root.add(importButton, GridConstraints(0, 4, 1, 1, ANCHOR_CENTER, FILL_HORIZONTAL, SIZEPOLICY_CAN_SHRINK or SIZEPOLICY_CAN_GROW, SIZEPOLICY_FIXED, null, null, null))
        root.add(resetButton, GridConstraints(0, 6, 1, 1, ANCHOR_CENTER, FILL_HORIZONTAL, SIZEPOLICY_CAN_SHRINK or SIZEPOLICY_CAN_GROW, SIZEPOLICY_FIXED, null, null, null))

        val tabbedPane = JTabbedPane()
        root.add(tabbedPane, GridConstraints(1, 0, 1, 7, ANCHOR_CENTER, FILL_BOTH, SIZEPOLICY_CAN_SHRINK or SIZEPOLICY_CAN_GROW, SIZEPOLICY_CAN_SHRINK or SIZEPOLICY_CAN_GROW, null, null, null))

        tabbedPane.addTab("Validators", JScrollPane(validators))
        validators.rowHeight = (validators.font.size * 1.5).toInt()
        validators.getDefaultEditor(String::class.java).addCellEditorListener(object: CellEditorListener {
            override fun editingCanceled(e: ChangeEvent?) {}
            override fun editingStopped(e: ChangeEvent?) = showValidatorPropertyErrorIfNeeded(e)
        })

        tabbedPane.addTab("Symbols", JScrollPane(symbols))
        symbols.rowHeight = (validators.font.size * 1.5).toInt()
    }

    internal fun showValidatorPropertyErrorIfNeeded(e: ChangeEvent?) {
        val text = (e?.source as CellEditor).cellEditorValue.toString()
        if (!isCorrectValidatorPropertiesFormat(text)) showValidatorPropertyError(text)
    }

    open internal fun showValidatorPropertyError(s: String) {
        Messages.showMessageDialog("Validator property must be in key=value format: " + s, "Invalid validator property format", Messages.getErrorIcon())
    }

    internal fun isCorrectValidatorPropertiesFormat(text: String): Boolean {
        return parseProperties(text) != null
    }

    open internal fun cloneConfigs() {
        provider.configs.forEach { e -> configs[e.key] = e.value.clone() }
    }

    val pane: JPanel
        get() {
            initLanguages()
            initTabs()
            initButtons()
            return root
        }

    open internal fun initButtons() {
        exportButton.addActionListener { a -> exportConfig() }
        importButton.addActionListener { a -> importConfig() }
        resetButton.addActionListener { a -> resetToDefaults() }
    }

    internal fun importConfig() {
        try {
            if (fileChooser.showOpenDialog(root) != APPROVE_OPTION) return
            config = configurationLoader.load(fileChooser.selectedFile)
            initTabs()
        } catch (e: RedPenException) {
            Messages.showMessageDialog("Cannot load: " + e.message, "RedPen", Messages.getErrorIcon())
        }
    }

    internal fun exportConfig() {
        try {
            if (fileChooser.showSaveDialog(root) != APPROVE_OPTION) return
            save()
            configurationExporter.export(config, FileOutputStream(fileChooser.selectedFile))
        } catch (e: IOException) {
            Messages.showMessageDialog("Cannot write to file: " + e.message, "RedPen", Messages.getErrorIcon())
        }

    }

    open internal fun initLanguages() {
        provider.configs.keys.forEach { k -> language.addItem(k) }
        language.selectedItem = provider.activeConfig.key
        language.addPopupMenuListener(object : PopupMenuListenerAdapter() {
            override fun popupMenuWillBecomeVisible(e: PopupMenuEvent?) {
                ensureTableEditorsStopped()
                applyChanges()
            }
        })
        language.addActionListener { a -> initTabs() }
    }

    open internal fun initTabs() {
        initValidators()
        initSymbols()
    }

    open internal fun initSymbols() {
        symbols.model = createSymbolsModel()
        symbols.columnModel.getColumn(0).minWidth = 250
        symbols.setDefaultEditor(Char::class.javaObjectType, SingleCharEditor())

        val symbolTable = config.symbolTable
        for (key in symbolTable.names) {
            val symbol = symbolTable.getSymbol(key)
            (symbols.model as DefaultTableModel).addRow(arrayOf(symbol.type.toString(), symbol.value, String(symbol.invalidChars), symbol.isNeedBeforeSpace, symbol.isNeedAfterSpace))
        }

        symbols.doLayout()
    }

    open internal fun initValidators() {
        validators.model = createValidatorsModel()
        validators.columnModel.getColumn(0).maxWidth = 20

        val validatorConfigs = config.validatorConfigs.groupByName()
        for (config in combinedValidatorConfigs()) {
            (validators.model as DefaultTableModel).addRow(arrayOf(validatorConfigs.containsKey(config.key), config.key, config.value.asString()))
        }

        validators.doLayout()
    }

    private fun ValidatorConfiguration.asString() = properties.entries.joinToString("; ")

    fun List<ValidatorConfiguration>.groupByName(): Map<String, ValidatorConfiguration> {
        return associateBy({ it.configurationName }, { it })
    }

    fun combinedValidatorConfigs() = provider.initialConfigs[config.key]!!.validatorConfigs.groupByName() + config.validatorConfigs.groupByName()

    open fun getEditedValidators(): List<ValidatorConfiguration> {
        val result = ArrayList<ValidatorConfiguration>()
        val model = validators.model
        val allConfigs = combinedValidatorConfigs()

        val editorText = ((validators.cellEditor as? DefaultCellEditor)?.component as? JTextField)?.text
        val editingRow = validators.editingRow
        fun editableValueAt(i: Int) = if (editingRow == i && editorText != null) editorText else model.getValueAt(i, 2) as String

        for (i in 0..model.rowCount-1) {
            if (model.getValueAt(i, 0) as Boolean) {
                val validator = allConfigs[model.getValueAt(i, 1)]!!.clone()
                validator.properties.clear()
                parseProperties(editableValueAt(i))?.forEach { validator.addProperty(it.key, it.value) }
                result.add(validator)
            }
        }
        return result
    }

    fun parseProperties(text: String): Map<String, String>? {
        return text.split(";\\s*".toRegex()).filter { it.isNotEmpty() }.map { it.split("=".toRegex(), 2) }.associate {
            if (it.size < 2 || it[0].isEmpty()) return null
            it[0].trim() to it[1]
        }
    }

    open fun getEditedSymbols(): List<Symbol> {
        val model = symbols.model

        val editorText = ((symbols.cellEditor as? DefaultCellEditor)?.component as? JTextField)?.text
        val editingRow = symbols.editingRow
        val editingColumn = symbols.editingColumn
        fun editableValueAt(i: Int, j: Int) = if (editorText != null && editingRow == i && editingColumn == j) editorText
                else model.getValueAt(i, j).toString()

        return (0..model.rowCount-1).map { i ->
            Symbol(SymbolType.valueOf(model.getValueAt(i, 0) as String), editableValueAt(i, 1)[0], editableValueAt(i, 2),
                   model.getValueAt(i, 3) as Boolean, model.getValueAt(i, 4) as Boolean)
        }
    }

    open internal fun applyValidatorsChanges() {
        val validators = config.validatorConfigs
        val editedValidators = getEditedValidators()
        validators.clear()
        validators.addAll(editedValidators)
    }

    open internal fun applySymbolsChanges() {
        val symbolTable = config.symbolTable
        getEditedSymbols().forEach { symbolTable.overrideSymbol(it) }
    }

    open internal fun ensureTableEditorsStopped() {
        if (validators.isEditing) validators.cellEditor.stopCellEditing()
        if (symbols.isEditing) symbols.cellEditor.stopCellEditing()
    }

    open internal fun applyChanges() {
        applyValidatorsChanges()
        applySymbolsChanges()
    }

    open fun save() {
        ensureTableEditorsStopped()
        applyChanges()
        provider.configs.putAll(configs)
        cloneConfigs()
    }

    open fun resetChanges() {
        cloneConfigs()
        initTabs()
    }

    fun resetToDefaults() {
        provider.initialConfigs.forEach { e -> configs[e.key] = e.value.clone() }
        initTabs()
    }

    internal fun createValidatorsModel(): DefaultTableModel {
        val model: DefaultTableModel = object : DefaultTableModel() {
            override fun getColumnClass(i: Int): Class<*> {
                return if (i == 0) Boolean::class.javaObjectType else String::class.java
            }

            override fun isCellEditable(row: Int, column: Int): Boolean {
                return column != 1
            }
        }
        model.addColumn("")
        model.addColumn("Name")
        model.addColumn("Properties")
        return model
    }

    internal fun createSymbolsModel(): DefaultTableModel {
        val model: DefaultTableModel = object : DefaultTableModel() {
            override fun getColumnClass(i: Int): Class<*> {
                return if (i == 1) Char::class.javaObjectType else if (i == 3 || i == 4) Boolean::class.javaObjectType else String::class.java
            }

            override fun isCellEditable(row: Int, column: Int): Boolean {
                return column != 0
            }
        }
        model.addColumn("Symbols")
        model.addColumn("Value")
        model.addColumn("Invalid chars")
        model.addColumn("Space before")
        model.addColumn("Space after")
        return model
    }

    open var config: Configuration
        get() = configs[language.selectedItem as String]!!
        set(config) {
            configs[config.key] = config
            language.selectedItem = config.key
            if (config.key != language.selectedItem) {
                provider += config
                language.addItem(config.key)
                language.selectedItem = config.key
            }
        }
}
